ipatests/test_xmlrpc/test_automember_plugin.py

```diff
@@ -229,6 +229,109 @@ def test_delete_with_nonexistent_hostgroup(self, automember_hostgroup,
             command()
 
+    def test_create_with_existent_group(self, automember_group, group1):
+        """ Try to add a rule that already exists """
+        group1.ensure_exists()
+        command = automember_group.make_create_command()
+        with raises_exact(errors.NotFound(
+                reason=u'Automember rule with name "%s" '
+                       u'already exists' % group1.cn)):
+            command()
+
+    def test_create_with_existent_hostgroup(self, automember_hostgroup,
+                                            hostgroup1):
+        """ Try to add a rule with non-existent group """
+        hostgroup1.ensure_exists()
+        command = automember_hostgroup.make_create_command()
+        with raises_exact(errors.NotFound(
+                reason=u'Automember rule with name "%s" '
+                       u'already exists' % hostgroup1.cn)):
+            command()
+
+
+class TestAutomemberFindNegativeBadgroup(XMLRPC_test):
+
+    def test_find_with_nonexistent_group(self, automember_group, group1):
+        """ Try to find a rule with non-existent group """
+        group1.find()
+        command = automember_group.make_find_command(type=u'group')
+        with raises_exact(errors.NotFound(
+                reason=u'0 rules matched')):
+            command()
+
+    def test_find_with_nonexistent_hostgroup(self, automember_hostgroup,
+                                             hostgroup1):
+        """ Try to find a rule with non-existent hostgroup """
+        hostgroup1.find()
+        command = automember_hostgroup.make_find_command(type=u'hostgroup')
+        with raises_exact(errors.NotFound(
+                reason=u'0 rules matched')):
+            command()
+
+
+class TestAutomemberFindNegativeBadtype(XMLRPC_test):
+
+    def test_find_with_group_invalidtype(self, automember_group, group1):
+        """ Try to find existing group rule with invalid type """
+        group1.find()
+        command = automember_group.make_find_command(type=u'badtype')
+        with raises_exact(errors.NotFound(
+                reason=u'invalid \'type\': must be one '
+                       u'of \'group\', \'hostgroup\'')):
+            command()
+
+    def test_find_with_hostgroup_invalidtype(self, automember_hostgroup,
+                                             hostgroup1):
+        """ Try to find existing hostgroup rule with invalid type """
+        hostgroup1.find()
+        command = automember_hostgroup.make_find_command(type=u'badtype')
+        with raises_exact(errors.NotFound(
+                reason=u'invalid \'type\': must be one '
+                       u'of \'group\', \'hostgroup\'')):
+            command()
+
+
+class TestAutomemberShowNegativeBadgroup(XMLRPC_test):
+
+    def test_show_with_nonexistent_group(self, automember_group, group1):
+        """ Try to show a rule with non-existent group """
+        group1.retrieve()
+        command = automember_group.make_retrieve_command(type=u'group')
```

Comment:
I cannot find specification for the `type` parameter, could you please point me out what implementation is it using (just method signature is fine)?